### PR TITLE
Regex bugfix for link. email, and telegram

### LINKED ISSUE #210: [PE-D][Tester D] Unhandled exception when invalid link is accessed
When an invalid link is added to a person:
random.link
and when the "project link" on person's card is pressed, system throws an unhandled exception.

-------------
Labels: `type.FunctionalityBug` `severity.Low`
original: laney0808/ped#10

### LINKED ISSUE #238: [PE-D][Tester D] Violation of unique identifier constraints on telegram handle
When a person with handle @bob is added, after which another person with handle @BOB is added, the second person can be added successfully.

However, Telegram handles are case-insensitive, so "BOB" is the same as "bob". In this case, duplicated unique identifier exists in the system.

-------------
Labels: `type.FunctionalityBug` `severity.Low`
original: laney0808/ped#16

### LINKED ISSUE #239: [PE-D][Tester D] Violation of unique identifier constriants by email
When a person with email johnD@example.com is added after another person with email johnd@example.com is added, the person is added successfully.

However, email is case-insensitive, which means both emails are the same. It would be best to factor in this restriction to prevent duplicated unique identifier.

-------------
Labels: `type.FunctionalityBug` `severity.Low`
original: laney0808/ped#17

### PR DESCRIPTION
**Pull request by tanjiajiajun**
Fixes #238
Fixes #239

FIxes #210

### CODE REVIEW

File: docs/UserGuide.md [hunk 1]
```diff
@@ -206,7 +209,7 @@ Format: `list [timetable]`
 
 Examples:
 * `list`
-* `list timetable` 
+* `list timetable`
 ![result for 'list timetable'](images/listTimetable.png)
 ### Editing a student: `edit`
```
Comment (marclamp) [inline docs/UserGuide.md:214]: empty line

```suggestion
![result for 'list timetable'](images/listTimetable.png)

### Editing a student: `edit`
```

File: src/main/java/educonnect/logic/parser/ParserUtil.java [hunk 2]
```diff
@@ -313,14 +321,21 @@ public static HashSet<DayOfWeek> parseDaysSpecified(String days) throws ParseExc
...
-        System.out.println("Valid link");
-        return new Link(trimmedLink);
+
+        if (m.group("scheme") == null) {
+            logger.info("No scheme found!"); //if url does not contain ftp/http/https
+            url = "https://" + url;
+        }
+
+
+        return new Link(url);
```
Comment (tanjiajiajun) [inline src/main/java/educonnect/logic/parser/ParserUtil.java:339]: @jeremiahlzz review

File: src/main/java/educonnect/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -143,13 +150,14 @@ public static TelegramHandle parseTelegramHandle(String telegramHandle) throws P
 
     /**
      * Parses a {@code String email} into an {@code Email}.
-     * Leading and trailing whitespaces will be trimmed.
+     * Leading and trail    ing whitespaces will be trimmed.
```
Comment (marclamp) [inline src/main/java/educonnect/logic/parser/ParserUtil.java:153]: typo?
```suggestion
     * Leading and trailing whitespaces will be trimmed.
```

File: src/main/java/educonnect/model/student/Link.java [hunk 4]
```diff
@@ -9,12 +9,24 @@
...
+    // 9) path e.g. /file/file
+    // 10) query string e.g. ?something
+    // 11) fragment e.g. #something
+    public static final String VALIDATION_REGEX = "^(?<scheme>(?:ftp|https?):\\/\\/)?+(?:(?<username>[a-zA-Z][\\w-.]"
+            + "{0,31})(?::(?<password>[!-~&&[^@$\\n\\r]]{6,255}))?@)?(?<subdomain>(?:[a-zA-Z0-9][a-zA-Z0-9-]{0,61}"
+            + "[a-zA-Z0-9]?\\.){0,127})(?<domain>[a-zA-Z0-9][a-zA-Z0-9-]{1,61}[a-zA-Z0-9])(?<tld>\\."
+            + "[a-zA-Z]{3,63})(?<cctld>\\.[a-zA-Z]{2})?(?<portnumber>:\\d{1,5})?(?<path>(?:\\/{1,2}[\\w-@.~()%]"
+            + "*)*)(?<querystring>\\?(?:[\\w-%]+=[\\w-?/:@.~!$&'()*+,;=%]+(?:&[\\w-%]+=[\\w-?/:@.~!$&'()*+,;=%]+)*)"
+            + "?)?(?<fragment>#[\\w-?/:@.~!$&'()*+,;=%]+)?$";
     //@@author
```
Comment (marclamp) [inline src/main/java/educonnect/model/student/Link.java:30]: extra author tag
```suggestion
```

**Review by marclamp — COMMENTED** [review]
Overall looks good, only some small minor typos

**Pull request merged**